### ISSUE TITLE
[problem] Failed to initialize NVML: Unknown Error

### ISSUE THREAD

**Issue by TymonLee** [Project-HAMi/HAMi#248]
### 1. Issue or feature description

Version info:
`Driver Version: 535.129.03   CUDA Version: 12.2`

I'm using k8s-vgpu-scheduler on our cluster. It worked for some time, but recently all the pods using vgpu failed to run `nvidia-smi`  with the error log shown below:

> root@gpu-pod-2:/# nvidia-smi
> [4pdvGPU Warn(28:140476096096064:util.c:135)]: NVML error at line 135: 1
> Failed to initialize NVML: Unknown Error

It is ok to run `nvidia-smi` in either the vgpu-device-plugin container or the node and gives the message:

```
root@10-3-8-171:/# nvidia-smi   
Mon Apr  8 08:51:14 2024       
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 535.129.03             Driver Version: 535.129.03   CUDA Version: 12.2     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  NVIDIA GeForce RTX 4090        Off | 00000000:24:00.0 Off |                  Off |
| 30%   31C    P8              17W / 450W |      2MiB / 24564MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
|   1  NVIDIA GeForce RTX 4090        Off | 00000000:25:00.0 Off |                  Off |
| 30%   30C    P8              14W / 450W |      2MiB / 24564MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
|   2  NVIDIA GeForce RTX 4090        Off | 00000000:26:00.0 Off |                  Off |
| 30%   33C    P2              51W / 450W |   8324MiB / 24564MiB |      4%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
|   3  NVIDIA GeForce RTX 4090        Off | 00000000:27:00.0 Off |                  Off |
| 30%   31C    P8              24W / 450W |      2MiB / 24564MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
|   4  NVIDIA GeForce RTX 4090        Off | 00000000:A3:00.0 Off |                  Off |
| 30%   35C    P2              43W / 450W |    903MiB / 24564MiB |      1%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
|   5  NVIDIA GeForce RTX 4090        Off | 00000000:A4:00.0 Off |                  Off |
| 30%   32C    P2              54W / 450W |   8148MiB / 24564MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
|   6  NVIDIA GeForce RTX 4090        Off | 00000000:A5:00.0 Off |                  Off |
| 30%   35C    P2              46W / 450W |   7009MiB / 24564MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
|   7  NVIDIA GeForce RTX 4090        Off | 00000000:A6:00.0 Off |                  Off |
| 30%   31C    P8              15W / 450W |      2MiB / 24564MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
                                                                                         
+---------------------------------------------------------------------------------------+
| Processes:                                                                            |
|  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
|        ID   ID                                                             Usage      |
|=======================================================================================|
|    2   N/A  N/A    914499      C   /tmp/starter/ln_binary/perception          7248MiB |
|    2   N/A  N/A    914506      C   /tmp/starter/ln_binary/lock_on_road        1058MiB |
|    4   N/A  N/A    916343    C+G   ...offscreen_playbag/offscreen_playbag      890MiB |
|    5   N/A  N/A    915279      C   /tmp/starter/ln_binary/perception          7078MiB |
|    5   N/A  N/A    915282      C   /tmp/starter/ln_binary/lock_on_road        1058MiB |
|    6   N/A  N/A    891941      C   /tmp/starter/ln_binary/perception          6992MiB |
+---------------------------------------------------------------------------------------+
```

### 3. Information to [attach](https://help.github.com/articles/file-attachments-on-issues-and-pull-requests/) (optional if deemed irrelevant)

Common error checking:
- [ ] The output of `nvidia-smi -a` on your host
- [ ] Your docker or containerd configuration file (e.g: `/etc/docker/daemon.json`)

containerd config:
```
version = 2
root = "/var/lib/docker/"
state = "/run/containerd"
oom_score = 0

[grpc]
  max_recv_message_size = 16777216
  max_send_message_size = 16777216

[debug]
  level = "info"

[metrics]
  address = ""
  grpc_histogram = false

[plugins]
  [plugins."io.containerd.snapshotter.v1.overlayfs"]
    mount_options = ['volatile']
  [plugins."io.containerd.grpc.v1.cri"]
    sandbox_image = "k8s.gcr.io/pause:3.6"
    max_container_log_line_size = -1
    [plugins."io.containerd.grpc.v1.cri".containerd]
      default_runtime_name = "nvidia"
      snapshotter = "overlayfs"
      [plugins."io.containerd.grpc.v1.cri".containerd.runtimes]
        [plugins."io.containerd.grpc.v1.cri".containerd.runtimes.nvidia]
          privileged_without_host_devices = false
          runtime_engine = ""
          runtime_root = ""
          runtime_type = "io.containerd.runc.v2"
          [plugins."io.containerd.grpc.v1.cri".containerd.runtimes.nvidia.options]
            BinaryName = "/usr/bin/nvidia-container-runtime"
            systemdCgroup = true
    [plugins."io.containerd.grpc.v1.cri".registry]
      [plugins."io.containerd.grpc.v1.cri".registry.mirrors]
        [plugins."io.containerd.grpc.v1.cri".registry.mirrors."master0:5000"]
          endpoint = ["http://10.16.10.163:5000"]
          [plugins."io.containerd.grpc.v1.cri".registry.configs."master0:5000".tls]
            insecure_skip_verify = true
        [plugins."io.containerd.grpc.v1.cri".registry.mirrors."docker.io"]
          endpoint = ["https://registry-1.docker.io"]
        [plugins."io.containerd.grpc.v1.cri".registry.mirrors."deeproute-simulation-registry-registry-vpc.cn-shenzhen.cr.aliyuncs.com"]
          endpoint = ["https://reg.deeproute.ai:443"]
```

- [ ] The vgpu-device-plugin container logs

no useful information

- [ ] The vgpu-scheduler container logs

no relevant info

- [ ] The kubelet logs on the node (e.g: `sudo journalctl -r -u kubelet`)

no relevant info

Additional information that might help better understand your environment and reproduce the bug:
- [ ] Docker version from `docker version`

containerd version=v1.6.24

- [ ] Docker command, image and tag used
- [ ] Kernel version from `uname -a`

> root@10-3-8-171:/# uname -a
> Linux 10-3-8-171.bm.pd.sz.deeproute.ai 5.15.0-92-generic #102~20.04.1-Ubuntu SMP Mon Jan 15 13:09:14 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux

- [ ] Any relevant kernel output lines from `dmesg`

no relevant info from dmesg

**Comment by haitwang-cloud** [comment]
@TymonLee I have a blog about how to fix this https://github.com/haitwang-cloud/blog/blob/main/GPU/failed-to-initialize-nvml-unknow-error.md PTAL.

**Comment by TymonLee** [comment]
> @TymonLee I have a blog about how to fix this https://github.com/haitwang-cloud/blog/blob/main/GPU/failed-to-initialize-nvml-unknow-error.md PTAL.

I tried to execute `nvidia-ctk system create-dev-char-symlinks --create-all` and then `systemctl daemon-reload`, but still not working, even if I have restarted the pod. 

I checked the default cgroup is systemd.

**Comment by TymonLee** [comment]
I noticed some "API mismatch" errors in the `dmesg ` logs as shown below. I'm not sure if this is relevant. Is there any client in this vgpu-device-plugin?

![image](https://github.com/Project-HAMi/HAMi/assets/23135265/f1c98cd8-33f7-4fe4-b280-2f8171162f17)

**Comment by haitwang-cloud** [comment]
Maybe u should try downgrading the driver version of your GPU? Then to monitor whether this issue still happen? Cause from what I know this vGPU device plugin only been under limit envs

**Comment by TymonLee** [comment]
> Maybe u should try downgrading the driver version of your GPU? Then to monitor whether this issue still happen? Cause from what I know this vGPU device plugin only been under limit envs

Finally I figured out the issue: 

I have 2 clusters and both use systemdCgroup, but only 1 cluster has this problem. So, I thought the issue may not because of the systemdCgroup.

But today, I found that **the Linux Kernels of these 2 clusters are different**. And I **set the systemdCgroup to false** and finally it worked! So only when you set systemdCgroup to true in specific kernel version will lead to this problem. Amazing... 

Anyway, thanks to your kindly reply, which really helped a lot!
